ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/interface.hh

```diff
@@ -0,0 +1,160 @@
+ * \f$ \hat{\Psi}(\la_1, \la_2, \la_3) \f$ and its first and second derivatives
+ * w.r.t the total principal stretches.
+ *
+ * \tparam DF Deviatoric function.
+ */
+template <Concepts::DeviatoricFunction DF>
+struct Deviatoric
+{
+  using ScalarType         = typename DF::ScalarType;
+  using PrincipalStretches = typename DF::PrincipalStretches;
+  using MaterialParameters = typename DF::MaterialParameters;
+
+  using DeviatoricFunction = DF;
+
+  using FirstDerivative  = typename DF::FirstDerivative;
+  using SecondDerivative = typename DF::SecondDerivative;
+
+  static constexpr int dim = 3;
+
+  using StressMatrix   = Eigen::Vector<ScalarType, dim>;
+  using MaterialTensor = Eigen::TensorFixedSize<ScalarType, Eigen::Sizes<3, 3, 3, 3>>;
+
+  [[nodiscard]] constexpr static std::string name() noexcept { return "Deviatoric function: " + DF::name(); }
+
+  Deviatoric(const DF df)
+      : deviatoricFunction_{df} {}
+
+  /**
+   * \brief Returns the material parameters stored in the deviatoric part of the material.
+   */
+  const MaterialParameters materialParameters() const { return deviatoricFunction_.materialParametersImpl(); }
+
+  /**
+   * \brief Returns the stored energy obtained from the deviatoric function.
+   *
+   * \param lambdas the principal stretches.
+   * \return ScalarType the energy.
+   */
+  ScalarType storedEnergy(const PrincipalStretches& lambda) const {
+    checkDuplicates(lambda);
+    return deviatoricFunction_.storedEnergyImpl(lambda);
+  };
+
+  /**
+   * \brief Returns the principal PK2 stresses obtained from the first derivative of the deviatoric function.
+   *
+   * \param lambda the principal stretches.
+   * \return StressMatrix the stresses in Cartesian coordinate system.
+   */
+  StressMatrix stresses(const PrincipalStretches& lambda) const {
+    checkDuplicates(lambda);
+    auto dWdLambda = deviatoricFunction_.firstDerivativeImpl(lambda);
+
+    // Compute the principal PK2 stresses by dividing by the stretches
+    StressMatrix S;
+    for (const auto k : dimensionRange())
+      S[k] = dWdLambda[k] / lambda[k];
+
+    return S;
+  }
+
+  /**
+   * \brief Returns the material tangent modulus obtained from the second derivative of the deviatoric function.
+   *
+   * \param lambda the principal stretches.
+   * \return MaterialTensor the tangentModuli in Cartesian coordinate system.
+   */
+  MaterialTensor tangentModuli(const PrincipalStretches& lambda) const {
+    checkDuplicates(lambda);
+    auto S  = stresses(lambda);
+    auto dS = deviatoricFunction_.secondDerivativeImpl(lambda);
+
+    auto L = MaterialTensor{};
+    L.setZero();
+
+    for (const auto i : dimensionRange())
+      for (const auto k : dimensionRange())
+        L(i, i, k, k) = 1.0 / (lambda(i) * lambda(k)) * dS(i, k);
+
+    for (const auto i : dimensionRange())
+      for (const auto k : dimensionRange())
+        if (i != k) {
+          if (Dune::FloatCmp::eq(lambda(i), lambda(k), 1e-8))
```

Comment:
Waht is the underlying reason for this if statement? MAybe this can be change to get rid of the tolerance and merge it with Line 112?

Reply:
I dont get the reasoning below at checkDuplicates and here. I also dont see where L'hopitals rule is used? Does this happen at the stresses? I'm confident that this can be fixed in a nice way. Also for Autodiff types they also would go through the simplified branch would they?

Reply:
![alt text](image.png)

Reply:
![grafik](https://github.com/user-attachments/assets/4200c58a-a7e0-4d5e-9c78-65f5496162c1)

Reply:
See Wriggers 2008 -  Nonlinear Finite Element Methods, p.77 for implementation details. 
The l'Hôpital rule is neccesary as the denominator -> 0 for $\lambda_i = \lambda_k$

Reply:
I am not sure anymore why Tarun added the checkDuplicate() method, but I think he had his reasons to make autodiff work

Reply:
If I remember correctly, if duplicates existed (for example in the undeformed state), using AutoDiff resulted in either nan or inf.. so it was a workaround which just throws error when AutoDiff is being used and duplicates existed.. the root underlying reason for AutoDiff giving here troubles was not further investigated. Both stresses and tangentModuli showed nan/inf I think even though the energy was correctly computed... We assumed AutoDiff doesn't utilise the l'hopital rule when such duplicates existed perhaps and somehow gives nans/infs..

Reply:
Ok now I got when this happens. Only when you use `autodiffMAT` and let it compute the tangent moduli it gives nan or inf?

Reply:
In the current interface where principal stretches are already computed from the outside I afraid we cannot do much to enable the autodiff, since in the autodif type from principle stretches there we already arriving in the stored energy function with nan.

Reply:
Yes exactly, this is all in prinicpal coordinate system. The transformation to cartesian coordinates happens in Hyperelastic class